// redux-falcor/lib/actions.js
'use strict';

exports.__esModule = true;
exports.retrievePath = retrievePath;
exports.retrievePaths = retrievePaths;
exports.retrieveValue = retrieveValue;
exports.setPath = setPath;
exports.setPaths = setPaths;
exports.callPath = callPath;
exports.clear = clear;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _falcorJsonGraph = require('falcor-json-graph');

var _tinyUuid = require('tiny-uuid');

var _tinyUuid2 = _interopRequireDefault(_tinyUuid);

var RETRIEVE_PATH = 'redux-falcor/falcor/RETRIEVE_PATH';
exports.RETRIEVE_PATH = RETRIEVE_PATH;
var RETRIEVE_PATHS = 'redux-falcor/falcor/RETRIEVE_PATHS';
exports.RETRIEVE_PATHS = RETRIEVE_PATHS;
var RETRIEVE_VALUE = 'redux-falcor/falcor/RETRIEVE_VALUE';
exports.RETRIEVE_VALUE = RETRIEVE_VALUE;
var SET_PATH = 'redux-falcor/falcor/SET_PATH';
exports.SET_PATH = SET_PATH;
var SET_PATHS = 'redux-falcor/falcor/SET_PATHS';
exports.SET_PATHS = SET_PATHS;
var CALL_PATH = 'redux-falcor/falcor/CALL_PATH';
exports.CALL_PATH = CALL_PATH;
var CLEAR = 'redux-falcor/falcor/CLEAR';

exports.CLEAR = CLEAR;

function retrievePath(path) {
  return {
    _id: _tinyUuid2['default'](),
    type: RETRIEVE_PATH,
    falcorPath: path,
    path: path
  };
}

function retrievePaths() {
  var args = Array.prototype.slice.call(arguments);

  return {
    _id: _tinyUuid2['default'](),
    type: RETRIEVE_PATHS,
    falcorPath: args,
    paths: args
  };
}

function retrieveValue(path) {
  return {
    _id: _tinyUuid2['default'](),
    type: RETRIEVE_VALUE,
    falcorPath: path,
    path: path
  };
}

function setPath(path, value) {
  return {
    _id: _tinyUuid2['default'](),
    type: SET_PATH,
    falcorPath: _falcorJsonGraph.pathValue(path, value),
    path: path,
    value: value
  };
}

function setPaths() {
  var args = Array.prototype.slice.call(arguments);

  return {
    _id: _tinyUuid2['default'](),
    type: SET_PATHS,
    falcorPath: args,
    paths: args
  };
}

function callPath(path, args, refPaths, thisPaths) {
  return {
    _id: _tinyUuid2['default'](),
    type: CALL_PATH,
    falcorPath: path,
    path: path,
    args: args,
    refPaths: refPaths,
    thisPaths: thisPaths
  };
}

function clear() {
  return { type: CLEAR };
}
