// redux-falcor/lib/index.js
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _actions = require('./actions');

var _middleware = require('./middleware');

var _middleware2 = _interopRequireDefault(_middleware);

var _reducer = require('./reducer');

var _reducer2 = _interopRequireDefault(_reducer);

exports.retrievePath = _actions.retrievePath;
exports.retrievePaths = _actions.retrievePaths;
exports.retrieveValue = _actions.retrieveValue;
exports.setPath = _actions.setPath;
exports.setPaths = _actions.setPaths;
exports.callPath = _actions.callPath;
exports.clear = _actions.clear;
exports.createFalcorMiddleware = _middleware2['default'];
exports.falcorReducer = _reducer2['default'];
exports.RETRIEVE_VALUE = _actions.RETRIEVE_VALUE;
exports.RETRIEVE_PATH = _actions.RETRIEVE_PATH;
exports.RETRIEVE_PATHS = _actions.RETRIEVE_PATHS;
exports.SET_PATH = _actions.SET_PATH;
exports.SET_PATHS = _actions.SET_PATHS;
exports.CALL_PATH = _actions.CALL_PATH;
exports.CLEAR = _actions.CLEAR;
